Ensure file name is valid when saving web page resources to disk

szewai opened the pull request:
#### 22dc25be180b6bfb18f6012b963885cc4e2797d1
<pre>
Ensure file name is valid when saving web page resources to disk
<a href="https://bugs.webkit.org/show_bug.cgi?id=262642">https://bugs.webkit.org/show_bug.cgi?id=262642</a>
rdar://116419303

Reviewed by Ryosuke Niwa.

We currently use the last component of URL as file name, and that could lead to file system error, because a valid URL
component may not be a valid file name. To fix that, we are going to generate valid file name based on last component
by:
1. encoding escape characters of file system.
2. ensuring length is smaller than 255 bytes.
3. appending sequence number when it&apos;s taken by other subresource file.
4. falling back to default file name when last component is empty.

Test: WebArchive.SaveResourcesValidFileName

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::generateValidFileName):
(WebCore::LegacyWebArchive::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/268970@main">https://commits.webkit.org/268970@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48d4cc330bfb88a8c52b9baacf555c818a18d151

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21176 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21720 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21399 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23898 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18249 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25515 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19330 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23378 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19918 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19205 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5076 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19788 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -62,6 +62,9 @@
 namespace WebCore {
 
+static constexpr unsigned maxFileNameSizeInBytes = 255;
+static constexpr char escapeCharactersForFileName[] = ":/";
```
Hm... do we really need to have a hard-coded list like this?

szewai replied:
🤔Could you be more specific about "this"? Do you think it should not be a list or should not be global?

szewai replied:
I tested more special names and it seems the other browsers do escape special characters by replacing them with ' ' or '_'. I previously said they don't as they don't encode the one I tested.  To match the escaping behavior, in the latest patch, I remove the hard-coded list and use `FileSystem::encodeForFileName` instead. Thanks for the suggestion. 
I also update the test to cover that.

szewai committed:
Ensure file name is valid when saving web page resources to disk
https://bugs.webkit.org/show_bug.cgi?id=262642
rdar://116419303

Reviewed by Ryosuke Niwa.

We currently use the last component of URL as file name, and that could lead to file system error, because a valid URL
component may not be a valid file name. To fix that, we are going to generate valid file name based on last component
by:
1. encoding escape characters of file system.
2. ensuring length is smaller than 255 bytes.
3. appending sequence number when it's taken by other subresource file.
4. falling back to default file name when last component is empty.

Test: WebArchive.SaveResourcesValidFileName

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::generateValidFileName):
(WebCore::LegacyWebArchive::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/268970@main
